### ISSUE TITLE
Capitalizing first word of story doesn't always work.

### ISSUE THREAD

**Issue by jonesmartins** [fabd/kanji-koohii#95]
When you write a story with the first character in lower case, the website turns it into upper case, but only if it's a letter.
I believe this is phpToolkit's fault, since I think the change would be made on line 148 of [kanji-koohii/lib/peer/StoriesPeer.php](https://github.com/fabd/kanji-koohii/blob/84888a649ac3c38ef9c15e1d339c267d61daec7b/lib/peer/StoriesPeer.php)
Maybe add a Regex match before capitalizing it, as you did on line 153 before adding a period to the story.

**Comment by fabd** [comment]
Hi jonesmartins

What is the expected behaviour? What other characters should it capitalize?

**Comment by jonesmartins** [comment]
Sorry, I forgot to add examples!

When we want to add markup to a story, we surround words with # for bold, and * for italic. 
Example:
For the word **plum**, my story is "plums grow on every tree".
Without markup, your software would turn it into:
> **Plum**s grow on every tree.

by turning the word "plum" bold. This is correct behaviour.

If the user writes the story *with* markup and the first character is not a letter such as:
`#plums# grow on *every tree*`
The result would be:
>**plums** grow on *every tree*. 

instead of
>**Plums** grow on *every tree*

This happens because # is the first character, so # gets capitalized instead of the first letter of "plums". Now I am not sure to which extent it'd be necessary to check for an actual letter, since a user could for some reason write a story such as "!%$%#$%!%$%#@something" and the software would need to ignore all the symbols before an actual letter/word and capitalize it, but Regex may do it fast enough.

**Comment by fabd** [comment]
Alright thanks. I labelled this "help wanted" because that's a low complexity issue for a contributor to start poking around the codebase. And it's not critical so I'll leave it there for a while in case someone wants to tackle it.

**Comment by jonesmartins** [comment]
Sure, I understand. :) 
Unfortunately I only know Python, so I wouldn't know an efficient way to do it in PHP.

**Comment by iveskins** [comment]
Maybe something like this
<?php
   preg_match("/(?P<start>[^A-Za-z]*)(?P<first_letter>[A-Za-z])(?P<end>.*$)/",
                       '?hello',
                       $matches);
   print_r($matches);
?>
in the mb_ucfirst function..

**Comment by iveskins** [comment]
Like this..
http://sandbox.onlinephpfunctions.com/code/14000214595fcce073ff984d920ae16de7a1c584


```
<?php
class phpToolkit
{
  public static function mb_ucfirst($string)
  {
    if (!function_exists('mb_ucfirst') && function_exists('mb_substr'))
    {
      $charset = "UTF-8";
      #$charset = sfConfig::get('sf_charset');
      preg_match("/(?P<start>[^A-Za-z]*)(?P<first_letter>[A-Za-z])(?P<end>.*$)/",
                       $string,
                       $matches);
      $start = $matches["start"];
      $first = $matches["first_letter"];
      $end = $matches["end"];
      $string = $start . mb_strtoupper($first, $charset) . $end;

    }
    else
    {
      throw new Exception(__METHOD__.': no mb_substr() support.');
    }
    return $string;
  }
};
$story = "*hello*";

$s = phpToolkit::mb_ucfirst($story);

print_r($s);

?>
```

**Comment by fabd** [comment]
Hey iveskins thanks I will look at this tomorrow afternoon hopefully.

Could you describe / document here simple test cases. ie. text entered in story box, what it is before, what it is expected to be formatted to.

**Comment by fabd** [comment]
Note regex are really slow and there may be much faster option. Perhaps looking for the first letter (a-z) and uppercase that.

One issue to keep in mind is that [getFormattedStory()](https://github.com/fabd/kanji-koohii/blob/84888a649ac3c38ef9c15e1d339c267d61daec7b/lib/peer/StoriesPeer.php#L142) is called for every story being rendered in the Shared Stories list, so performance is a concern.

**Comment by fabd** [comment]
I haven't looked in detail yet (supposed to be taking a break today).. but at a quick glance the [code](https://github.com/fabd/kanji-koohii/blob/84888a649ac3c38ef9c15e1d339c267d61daec7b/lib/peer/StoriesPeer.php#L148) here seems like it needs a change of logic rather than updating the mb_ucfirst function.

mb_ucffirst() doesn't care about words, it's a generic function that is filling in a potentially missing native php function in the mb_* group (multibyte character string functions).

So at a quick glance mind you , I think it's a problem of logic here. That line of code just says, ok let's uppercase the first character.

What jonesmartins points out is a plain bug in the logic, since first character of a word can be preceded by the formatting characters.

**Comment by fabd** [comment]
That is correct. See [ucfirst documentation](http://php.net/manual/en/function.ucfirst.php).

> Returns a string with the first character of str capitalized, **if that character is alphabetic.**

So mb_ucfirst is a multibyte version of that same behaviour, it will leave eg. `#keyword# bla bla` uncapitalized, so should the mb_ucfirst() function.

**Comment by iveskins** [comment]
Ok, how about checking the first character of the story string in the getFormattedStory function against any array of all the markdown characters that could be used (#,*,{ .. and what else? I cant find any help page about how to use the markdown for writing stories :( ) .. then if the character is not in the array, passing it to mb_ucfirst .. and if it is a special character passing it to a slower mb_ucfirst_regex... which wont be used often so shouldn't impact performance
Example here
http://sandbox.onlinephpfunctions.com/code/ffd5e8acce84b7fe7c252935ee3bc3436279c178

**Comment by fabd** [comment]
I will fix this myself. My bad, I should tag "help wanted" only things I can't do myself in an hour or less, otherwise we all waste time. Just learning the ropes.

**Comment by iveskins** [comment]
Thank you for your continued hard work on this project. It is helping me everyday. I wish I could contribute more. This issue is not so simple that it can be solved by a PHP novice it seems.

**Comment by fabd** [comment]
No problem :)
